python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -945,37 +948,44 @@ def pack(
         self,
         metadata_dict: Mapping[str, MetadataValue],
         whitelist_map: WhitelistMap,
-        descent_path: str,
+        context: RecursiveDescentContext,
     ) -> Sequence[Mapping[str, Any]]:
         return [
             {
                 "__class__": "EventMetadataEntry",
                 "label": k,
                 # MetadataValue itself can't inherit from NamedTuple and so isn't a PackableValue,
                 # but one of its subclasses will always be returned here.
-                "entry_data": pack_value(v, whitelist_map, descent_path),  # type: ignore
+                "entry_data": pack_value(v, whitelist_map, context),  # type: ignore
                 "description": None,
             }
             for k, v in metadata_dict.items()
         ]
 
     def unpack(
         self,
-        metadata_entries: List[Mapping[str, Any]],
+        metadata_entries: Union[List[Mapping[str, Any]], List["MetadataEntry"]],
         whitelist_map: WhitelistMap,
-        descent_path: str,
+        context: UnpackContext,
     ) -> Mapping[str, MetadataValue]:
-        return {
-            e["label"]: unpack_value(
-                e["entry_data"],
-                # MetadataValue itself can't inherit from NamedTuple and so isn't a PackableValue,
-                # but one of its subclasses will always be returned here.
-                as_type=MetadataValue,  # type: ignore
-                whitelist_map=whitelist_map,
-                descent_path=descent_path,
-            )
-            for e in metadata_entries
-        }
+        if isinstance(context, RecursiveDescentContext):
+            metadata_entries = cast(List[Mapping[str, Any]], metadata_entries)
+            return {
+                e["label"]: unpack_value(
+                    e["entry_data"],
+                    # MetadataValue itself can't inherit from NamedTuple and so isn't a PackableValue,
+                    # but one of its subclasses will always be returned here.
+                    as_type=MetadataValue,  # type: ignore
+                    whitelist_map=whitelist_map,
+                    context=context,
+                )
+                for e in metadata_entries
+            }
+        elif isinstance(context, BottomUpUnpackContext):
+            metadata_entries = cast(List[MetadataEntry], metadata_entries)
+            return {e.label: e.entry_data for e in metadata_entries}
+        else:
+            check.failed(f"unexpected serdes context {context}")
```

Comment:
Complexity unleashed by this change is having two modes of unpacking. One bottom up when we are deserializing from a string and the other recursive descent when we are unpacking an already inflated structure. 

We can handle this forked behavior in different ways, such as having two `unpack` methods if we think thats better.

Reply:
what's the type of the 'inflated structure' being referred to here? Some internal json object representation?

I think i would find this clearer fwiw if they're not sharing a lot of code - unpack_string vs. unpack_<answer to question above>

Reply:
> what's the type of the 'inflated structure' being referred to here? Some internal json object representation?

The default `json.loads` output for JSON object is `dict` - so that tree of `dict` / `list` containers that we recurse over in search of the special `dict` entries containing the `__class__` / `__enum__` keys that trigger special unpack behavior 

> I think i would find this clearer fwiw if they're not sharing a lot of code - unpack_string vs. unpack_

👍  curious to hear what others think too- can take a pass at harder bifurcation of paths

Reply:
> Is there a way to prevent deserialization into MetadataEntry by this bottom-up process?

I believe the way this would be handled once `MetadataEntry` is removed is that you would get a `UnknownSerdesValue` which you would then process the `.value` of like we do in `[1]`

Reply:
> recursive descent when we are unpacking an already inflated structure

I'm still not clear why we would need the `RecursiveDescentContext` in an unpack... when do we have the inflated structure?

Reply:
It's exclusively for when you are calling `unpack_value` directly (i.e. not going through `json.loads`). @alangenfeld and I talked offline about it and think it's possible to change `unpack_value` to just go bottom up as well, which eliminates the need for two kinds of contexts.